File: Notes/Computer Science/Algorithms/Meta/Sebastien Rubens Meta Puzzles Solutions/cs/l1_rotary_lock1.cs
// meta_puzzles by Sebastien Rubens
//
// Please go to https://github.com/seb-pg/meta_puzzles/README.md
// for more information
//
// To the extent possible under law, the person who associated CC0 with
// meta_puzzles has waived all copyright and related or neighboring rights
// to meta_puzzles.
//
// You should have received a copy of the CC0 legalcode along with this
// work.  If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.

using System;
using System.Collections.Generic;

namespace l1_rotary_lock1
{

class Solution {

    public long getMinCodeEntryTime(int N, int M, int[] C) {
        // https://www.metacareers.com/profile/coding_puzzles/?puzzle=990060915068194
        // Constraints
        //      3 ≤ N ≤ 50,000,000      N is the number of integers
        //      1 ≤ M ≤ 1,000           M is the number of locks
        //      1 ≤ Ci ≤ N              Ci is the combination
        // Complexity: O(M)

        // The following is O(N)
        int pos = 1;
        long nb = 0;
        foreach (var target in C)
        {
            var positive_move = (target - pos) % N;  // positive move
            positive_move = positive_move < 0 ? positive_move + N : positive_move;  // modulo can be negative in C++
            var negative_move = N - positive_move;
            nb += Math.Min(positive_move, negative_move);
            pos = target;
        }
        return nb;
    }

    class Args : test_all.Res<long>
    {
        public int N;
        public int[] C;
    }

    public static int tests()
    {
        var s = new Solution();

        Func<Args, long> wrapper = (Args p) => s.getMinCodeEntryTime(p.N, p.C.Length, p.C);

        var args_list = new List<Args> {
            new Args { N=3, C=new int[] { 1, 2, 3 }, res=2 },
            new Args { N=10, C=new int[] { 9, 4, 4, 8 }, res=11 },
        };

        return test_all.TestAll.run_all_tests("l1_rotary_lock1", args_list, wrapper);
    }

}

}

File: Notes/Computer Science/Algorithms/Meta/Sebastien Rubens Meta Puzzles Solutions/rs/src/l1_rotary_lock1.rs
// meta_puzzles by Sebastien Rubens
//
// Please go to https://github.com/seb-pg/meta_puzzles/README.md
// for more information
//
// To the extent possible under law, the person who associated CC0 with
// meta_puzzles has waived all copyright and related or neighboring rights
// to meta_puzzles.
//
// You should have received a copy of the CC0 legalcode along with this
// work.  If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.

#![allow(non_snake_case)]

pub fn getMinCodeEntryTime(N: i32, _M: i32, C: &Vec<i32>) -> i64 {
    use std::cmp;

    let mut pos: i32 = 1;
    let mut nb: i64 = 0;
    for &target in C {
        let mut positive_move = (target - pos) % N;  // positive move
        positive_move = if positive_move < 0 { positive_move + N } else { positive_move};  // modulo can be negative in C++
        let negative_move = N - positive_move;
        nb += cmp::min(positive_move, negative_move) as i64;
        pos = target;
    }
    return nb;
}


type RetType = i64;

struct Args
{
    N: i32,
    C: Vec<i32>,
    res: RetType,
}

impl super::Result<RetType> for Args {
    fn get_result(&self) -> RetType
    {
        return self.res;
    }
}

pub fn tests() -> u32
{
    let wrapper = |p: &Args| -> RetType { getMinCodeEntryTime(p.N, p.C.len() as i32, &p.C) };

    let args_list : Vec<Args> = vec![
        Args{ N: 3, C: vec![ 1, 2, 3 ], res: 2 },
        Args{ N: 10, C: vec![ 9, 4, 4, 8 ], res: 11 },
    ];

    return super::run_all_tests("l1_rotary_lock1", args_list, wrapper, Option::None);
}
